docs/design/security_context.md

```diff
@@ -0,0 +1,158 @@
+should be able to be uniquely identified by their Unix UID, such that a chain of  ownership can be established. 
+Processes in pods will need to have consistent UID/GID/SELinux category labels in order to access shared disks.
+
+## Constraints and Assumptions
+* It is out of the scope of this document to prescribe a specific set 
+  of constraints to isolate containers from their host. Different use cases need different
+  settings.
+* The concept of a security context should not be tied to a particular security mechanism or platform 
+  (ie. SELinux, AppArmor)
+* Applying a different security context to a scope (namespace or pod) requires a solution such as the one proposed for
+  [service accounts](https://github.com/GoogleCloudPlatform/kubernetes/pull/2297).
+
+## Use Cases
+
+In order of increasing complexity, following are example use cases that would 
+be addressed with security contexts:
+
+1.  Kubernetes is used to run a single cloud application. In order to protect
+    nodes from containers:
+    * All containers run as a single non-root user
+    * Privileged containers are disabled
+    * All containers run with a particular MCS label 
+    * Kernel capabilities like CHOWN and MKNOD are removed from containers
+    
+2.  Just like case #1, except that I have more than one application running on
+    the Kubernetes cluster.
+    * Each application is run in its own namespace to avoid name collisions
+    * For each application a different uid and MCS label is used
+    
+3.  Kubernetes is used as the base for a PAAS with 
+    multiple projects, each project represented by a namespace. 
+    * Each namespace is associated with a range of uids/gids on the node that
+      are mapped to uids/gids on containers using linux user namespaces. 
+    * Certain pods in each namespace have special privileges to perform system
+      actions such as talking back to the server for deployment, run docker
+      builds, etc.
+    * External NFS storage is assigned to each namespace and permissions set
+      using the range of uids/gids assigned to that namespace. 
+
+## Proposed Design
+
+### Overview
+A *security context* consists of a set of constraints that determine how a container
+is secured before getting created and run. It has a 1:1 correspondence to a
+[service account](https://github.com/GoogleCloudPlatform/kubernetes/pull/2297). A *security context provider* is passed to the Kubelet so it can have a chance
+to mutate Docker API calls in order to apply the security context.
+
+It is recommended that this design be implemented in two phases:
+
+1.  Implement the security context provider extension point in the Kubelet 
+    so that a default security context can be applied on container run and creation.
+2.  Implement a security context structure that is part of a service account. The
+    default context provider can then be used to apply a security context based
+    on the service account associated with the pod.
```

Comment:
Correct - I believe that a reasonable default would be that each security context on the master (a service account that is the "default" for the namespace?) would get a UID allocated to it that no other security context would get.  An administrator would then later be able to assign complementary UIDs across namespaces if needed.  In the future, there could be additional security contexts that grant access to shared resources.

----- Original Message -----
> > +
> > +## Proposed Design
> > +
> > +### Overview
> > +A *security context* consists of a set of constraints that determine how a
> > container
> > +is secured before getting created and run. It has a 1:1 correspondence to
> > a
> > +[service
> > account](https://github.com/GoogleCloudPlatform/kubernetes/pull/2297). A
> > *security context provider* is passed to the Kubelet so it can have a
> > chance
> > +to mutate Docker API calls in order to apply the security context.
> > +
> > +It is recommended that this design be implemented in two phases:
> > +
> > +1.  Implement the security context provider extension point in the Kubelet
> > +    so that a default security context can be applied on container run and
> > creation.
> > +2.  Implement a security context structure that is part of a service
> > account. The
> > +    default context provider can then be used to apply a security context
> > based
> > +    on the service account associated with the pod.
> 
> If pods on different nodes are accessing shared storage, their UIDs need to
> be be unique across nodes.  So, their uids need to be allocated either
> statically to nodes at node join time, or dynamically to pods at bind time,
> by some cluster level thing.  Thoughts?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3910/files#r23849135